documentation/sphinx/source/developer-guide.rst

```diff
@@ -823,3 +823,133 @@ Loading data is a common task in any database. Loading data in FoundationDB will
 * Use multiple processes loading in parallel if a single one is CPU-bound.
 
 Using these techniques, our cluster of 24 nodes and 48 SSDs loads about 3 billion (100 byte) key-value pairs per hour.
+
+Implementation Details
+======================
+
+These following sections go into some of the gritty details of FoundationDB. Most users don't need to read or understand this in order to use FoundationDB efficiently.
+
+How FoundationDB detects Conflicts
+----------------------------------
+
+As written above, FoundationDB implements serializable transactions with external consistency. The underlying algorithm uses multi-version concurrency control. At commit time, each transaction is checked for read-write conflicts.
+
+Conceptually this algorithm is quite simple. Each transaction will get a read version assigned when it issues the first read or before it tries to commit. All reads that happen during that transaction will be read as of that version. Writes will go into a local cache and will be sent to FoundationDB during commit time. The transaction can successfully commit if it is conflict free; it will then get a commit-version assigned. A transaction is conflict free if and only if there have been been no writes to any key that was read by that transaction between the time the transaction started and the commit time. This is true if there was no transaction with a commit version larger than our read version but smaller than our commit version that wrote to any of the key that we wrote.
+
+This form of conflict detection, while simple, can often be confusing for people who are familiar with databases that check for write-write conflicts.
+
+Some interesting properties of FoundationDB transactions are:
+
+* FoundationDB transactions are optimistic: we never block on reads or writes (there are now locks), instead we abort transactions at commit time.
+* Read-only transactions will never conflict and never cause conflicts with other transactions.
```

Comment:
I would add, just to be extra explicit, to the "Read-only transactions will never":
this applies both to snapshot transactions as well as general read-write transactions that only perform reads.

Reply:
[The documentation](https://apple.github.io/foundationdb/developer-guide.html#snapshot-isolation) never talks about snapshot transactions but only about snapshot reads. And I would like to keep it that way.

A snapshot transaction typically is a transaction that only checks for write-write conflicts. This kind of transaction doesn't exist in fdb (we could easily build them manually, but there's no option to create a snapshot transaction). 

Doesn't the third point address what you mean? There I wrote this pretty explicitely:

```
For read-write transactions: A read will never cause any other transaction to be aborted
```

Reply:
Could you update https://apple.github.io/foundationdb/kv-architecture.html#resolvers:
"A transaction conflicts if it reads a key that has been written between the transaction’s read version and commit version."

and https://apple.github.io/foundationdb/developer-guide.html#transactions-in-foundationdb:
"(i.e. that no value read by a transaction has been modified by another transaction since the read occurred)"

Both of these places appear to be defining conflict at least somewhat rigorously, and they fail to mention that a transaction that has no write conflict range will never conflict.  I think this is hinted in the documentation when it says "Transactions that are not read-only must be committed, and the commit will not succeed until the transaction is fully (redundantly) durable.", but this does not necessarily imply that there will be no conflict even if the user tries to commit a read-only transaction.

Reply:
Good catch, I addressed this in both places.